ToricVarieties: Faster Stanley-Reisner ideal, (default) coefficient ring, names for homogeneous variables and fmpz-valued rays from polymake

HereAround opened the pull request:
This PR aims towards resolving the discussion in https://github.com/oscar-system/Oscar.jl/issues/932. In the current form, I find this draft NOT satisfying. I am certain that other people will find more issues. Here is what I see at this point:

As @fieker reported, there seem to be an issue with the following function whenever we have large/huge rays:

> function map_from_character_to_principal_divisors(v::AbstractNormalToricVariety)
mat = transpose(Matrix{Int}(Polymake.common.primitive(pm_object(v).RAYS)))
matrix = AbstractAlgebra.matrix(ZZ, mat)
return hom(character_lattice(v), torusinvariant_divisor_group(v), matrix)
end

Specifically, the conversion into `Matrix{Int}` can then fail. Generally speaking, it should be possible to replace this line via `rays(fan(v))`. This is currently not possible, because the return value is then `Polymake.Rationals`, while one should expect to have `fmpz`. As a temporary fix, I added a corresponding function in https://github.com/oscar-system/Oscar.jl/pull/939/commits/01d8f488d8668282d3249b699995d902b33542d5. @lkastner and @micjoswig: What are your thoughts on this?

To compute the Stanley-Reisner ideal via a simplicial complex, we can first focus on projective varieties, so that the fan is the normal fan of a polyhedron `P`. We do not have a function which computes this polyhedron yet. Is there functionality for this somewhere? If yes, I could not (yet) find it. A temporary fix is to cache the polyhedron, if the variety is created from one: https://github.com/oscar-system/Oscar.jl/pull/939/commits/4abd4640bf300f09ed859936d68a63f52a412d10

Next, one could hope to call `stanley_reisner_ideal(S,K)` , with S the graded (!) Cox ring and K the simplicial complex in question. This is not possible because currently only non-graded rings over `Z` are accepted as input:

> function stanley_reisner_ideal(R::FmpzMPolyRing, K::SimplicialComplex)
n = nvertices(K)
return ideal([ R([1], [_characteristic_vector(f,n)]) for f in minimal_nonfaces(K) ])
end

In particular, the expression `ideal([ R([1], [_characteristic_vector(f,n)]) for f in minimal_nonfaces(K) ])` cannot seem to be evaluated for a graded ring `R`. Ring experts, please comment.

The following change would work, but is probably not robust enough (see https://github.com/oscar-system/Oscar.jl/pull/939/commits/10cd389114e1a44312eebb3c930d7b41d71d546b):

> function stanley_reisner_ideal(R::MPolyRing, K::SimplicialComplex)
n = nvertices(K)
collections = [findall(x->x==1, _characteristic_vector(f,n)) for f in minimal_nonfaces(K)]
vars = Hecke.gens(R)
generators = [prod(vars[I]) for I in collections]
return ideal(generators)
end

Finally, I change the homogeneous coordinate from `x[i]` to `xi`. This looks nicer, but has other drawbacks (e.g. as they cannot be generated programmatically as @fieker pointed out). The relevant commits are:
- https://github.com/oscar-system/Oscar.jl/pull/939/commits/f5d0c4eef0601655467d999972a811439d5cdb31: Change `x[i]` to `xi`.
- https://github.com/oscar-system/Oscar.jl/pull/939/commits/328646a9e07f81b5d3d8c9ce3b321598573e07e9: An attempt to label the blowup coordinate as `ei`. (Not sufficiently robust at this point, in my opinion.)
- https://github.com/oscar-system/Oscar.jl/pull/939/commits/17cf7a1e9f34efe90803f3295dd0714d20993139: A method to allow the user to choose variable names for the Cox ring as he/she pleases. @fieker: I believe you had something more global in mind.

Last but not least, it would appear that `Z`, `Q` as base rings have each advantages/disadvantages. No resolution for this yet.

Finally, to rebase more easily, this PR contains the changes of https://github.com/oscar-system/Oscar.jl/pull/929 and https://github.com/oscar-system/Oscar.jl/pull/940, which I hope are both a lot less critical and can be merged very soon.

HereAround committed:
ToricVarieties: Change homogeneous variables of Cox ring: x[k] -> xk

HereAround committed:
ToricVarieties: Change order of homogeneous variables for del Pezzo surfaces

HereAround committed:
ToricVarieties: Use matrix(ZZ, rays(fan(v))) and nrays to process fans of ToricVarieties
-> Fix issue reported by @fieker in https://github.com/oscar-system/Oscar.jl/issues/932

HereAround committed:
ToricVarieties: Cache polyhedron of ToricVariety upon construction

HereAround committed:
ToricVarieties: Ask for user input for blowup_on_ith_torus_orbit

thofma commented on src/ToricVarieties/NormalToricVarieties/attributes.jl:
````diff
@@ -64,17 +64,49 @@ julia> p2 = toric_projective_space(2)
+  u3 -> [0 0 1], MPolyElem_dec{fmpq, fmpq_mpoly}[u1, u2, u3])
+```
+"""
+function set_cox_ring(v::AbstractNormalToricVariety, var_list::Vector{String})
````
```suggestion
function set_cox_ring!(v::AbstractNormalToricVariety, var_list::Vector{String})
```

HereAround replied:
Thank you!

HereAround commented:
@thofma I just force pushed and thereby accidentally removed the commit with your suggested change. I amended it to the latest commit, but the "co-authored by" text is gone. I could try to add this by hand if you care. Sorry about that.

thofma commented:
No worries :)

fieker commented:
You could replace `FmpzMPolyRing` by 'MPolyRing{fmpz}', try `supertype(FmpzMPolyRing)` followed by `subtypes(ans)`. This will cover both graded and non-graded rings.
The QQ vs ZZ: it really depends what you want to do with it. Any classical commutative algebra (Groebner and friends) and you're sunk when using ZZ. Combinatorcs and easyly moving to FF_p, then ZZ is better.
The fact that the generators of the ideal are in ZZ[x] is irrelevant as this is true for all ideals...

fieker commented:
On Mon, Jan 03, 2022 at 08:44:13AM -0800, Martin Bies wrote:
> @HereAround commented on this pull request.
> 
> 
> 
> > +
> +# Examples
> +```jdoctest
> +julia> p2 = NormalToricVariety(normal_fan(Oscar.simplex(2)))
> +A normal, non-affine, smooth, projective, gorenstein, q-gorenstein, fano, 2-dimensional toric variety without torusfactor
> +
> +julia> set_cox_ring(p2, "u1, u2, u3")
> +
> +julia> cox_ring(p2, "u1, u2, u3")
> +(Multivariate Polynomial Ring in u1, u2, u3 over Rational Field graded by 
> +  u1 -> [0 0 1]
> +  u2 -> [0 0 1]
> +  u3 -> [0 0 1], MPolyElem_dec{fmpq, fmpq_mpoly}[u1, u2, u3])
> +```
> +"""
> +function set_cox_ring(v::AbstractNormalToricVariety, var_list::Vector{String})
> 
> Thank you!

Cannot work, "u1, u2, u3" is of wrong type
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/oscar-system/Oscar.jl/pull/939#discussion_r777586607
> You are receiving this because you are subscribed to this thread.
> 
> Message ID: ***@***.***>

HereAround commented:
> On Mon, Jan 03, 2022 at 08:44:13AM -0800, Martin Bies wrote: @HereAround commented on this pull request. > + +# Examples +```jdoctest +julia> p2 = NormalToricVariety(normal_fan(Oscar.simplex(2))) +A normal, non-affine, smooth, projective, gorenstein, q-gorenstein, fano, 2-dimensional toric variety without torusfactor + +julia> set_cox_ring(p2, "u1, u2, u3") + +julia> cox_ring(p2, "u1, u2, u3") +(Multivariate Polynomial Ring in u1, u2, u3 over Rational Field graded by + u1 -> [0 0 1] + u2 -> [0 0 1] + u3 -> [0 0 1], MPolyElem_dec{fmpq, fmpq_mpoly}[u1, u2, u3]) +``` +""" +function set_cox_ring(v::AbstractNormalToricVariety, var_list::Vector{String}) Thank you!
> Cannot work, "u1, u2, u3" is of wrong type
> […](#)
> -- Reply to this email directly or view it on GitHub: [#939 (comment)](https://github.com/oscar-system/Oscar.jl/pull/939#discussion_r777586607) You are receiving this because you are subscribed to this thread. Message ID: ***@***.***>

Thank you!

HereAround commented:
> You could replace `FmpzMPolyRing` by 'MPolyRing{fmpz}', try `supertype(FmpzMPolyRing)` followed by `subtypes(ans)`. This will cover both graded and non-graded rings. The QQ vs ZZ: it really depends what you want to do with it. Any classical commutative algebra (Groebner and friends) and you're sunk when using ZZ. Combinatorcs and easyly moving to FF_p, then ZZ is better. The fact that the generators of the ideal are in ZZ[x] is irrelevant as this is true for all ideals...

As already proposed by @micjoswig , it would seem that we should simply allow an additional argument, with which one can specify if the variety is to be constructed over ZZ, QQ, F_p... I propose QQ as default.

fieker commented:
Your missing the point, my suggestion allows gradings...


On Tue, 4 Jan 2022, 18:05 Martin Bies, ***@***.***> wrote:

> You could replace FmpzMPolyRing by 'MPolyRing{fmpz}', try
> supertype(FmpzMPolyRing) followed by subtypes(ans). This will cover both
> graded and non-graded rings. The QQ vs ZZ: it really depends what you want
> to do with it. Any classical commutative algebra (Groebner and friends) and
> you're sunk when using ZZ. Combinatorcs and easyly moving to FF_p, then ZZ
> is better. The fact that the generators of the ideal are in ZZ[x] is
> irrelevant as this is true for all ideals...
>
> As already proposed by @micjoswig <https://github.com/micjoswig> , it
> would seem that we should simply allow an additional argument, with which
> one can specify if the variety is to be constructed over ZZ, QQ, F_p... I
> propose QQ as default.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/oscar-system/Oscar.jl/pull/939#issuecomment-1005005748>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA36CV4MBLTBLLEA3RF3EALUUMSD5ANCNFSM5LFRGOQQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

micjoswig commented:
> The fact that the generators of the ideal are in ZZ[x] is irrelevant as this is true for all ideals...

Well, truth is, the quotient is usually more important than the ideal; so there is a difference.  I am absolutely in favor of picking QQ coeffs for (the cohomology rings of) the toric varieties. As far as the Cox ring is concerned, I'd suggest QQ[x1,x2,...] as the default.  Of course, we need options for gradings and possibly other rings.

The interface for the simplicial complexes should be similar. While I have a slight preference to make ZZ[x1,x2,...] the default ring, I do not insist.  An argument in favor of QQ[x1,x2,...] here, too, would be that this would make it less likely for a user to do something stupid (with respect to computational resources). At any rate, here I do need (at least) the option to specify, e.g., ZZ[x1,x2,...] when I mean it.

benlorenz commented:
> As @fieker reported, there seem to be an issue with the following function whenever we have large/huge rays:
> 
> > function map_from_character_to_principal_divisors(v::AbstractNormalToricVariety)
> > mat = transpose(Matrix{Int}(Polymake.common.primitive(pm_object(v).RAYS)))
> > matrix = AbstractAlgebra.matrix(ZZ, mat)
> > return hom(character_lattice(v), torusinvariant_divisor_group(v), matrix)
> > end
> 
> Specifically, the conversion into `Matrix{Int}` can then fail. Generally speaking, it should be possible to replace this line via `rays(fan(v))`. This is currently not possible, because the return value is then `Polymake.Rationals`, while one should expect to have `fmpz`. As a temporary fix, I added a corresponding function in [01d8f48](https://github.com/oscar-system/Oscar.jl/commit/01d8f488d8668282d3249b699995d902b33542d5). @lkastner and @micjoswig: What are your thoughts on this?

The rays of a polyhedral fan or cone are usually considered in QQ (and not in ZZ) as this is required for any convex hull computations, and to retrieve them as a fmpq matrix you can use:
```julia
m = vector_matrix(rays(fan(v)))
```
There is probably some way to eliminate the denominators in a fmpq matrix to get a fmpz matrix (instead of using `Polymake.common.primitive`) ?

_Edit:_ We could probably also add something like `vector_matrix(fmpz, rays(fan(v)))` to prescribe the return type.

HereAround committed:
ToricVarieties: Print coefficient ring in Base.show

HereAround commented:
I have just updated this PR and hope that it addresses all points and issues made in https://github.com/oscar-system/Oscar.jl/issues/932. Hence, I marked it ready for review. Let me summarize the changes:

- Changes to the homogeneous coordinates of the Cox ring: https://github.com/oscar-system/Oscar.jl/pull/939/commits/c2bbf878ea760261a139e67f4752a14ae1e60c8e, https://github.com/oscar-system/Oscar.jl/pull/939/commits/e3c66709523a93b1f7899cc363fefafe748681d6, https://github.com/oscar-system/Oscar.jl/pull/939/commits/c2bbf878ea760261a139e67f4752a14ae1e60c8e, https://github.com/oscar-system/Oscar.jl/pull/939/commits/cb1c6127d35a94595ca05c5417930b640405a22d. In particular, I change them from `x[k]` to `xk` and ask for user input to name the exceptional coordinate of a blowup, as suggested by @fieker.
- Support for different coefficient rings as requested by @micjoswig : https://github.com/oscar-system/Oscar.jl/pull/939/commits/05adcef508ec6eca4c2fbb1edca4ad40eddc7e61. (I was originally tempted to name this the `base_ring`, but have decided for `coefficient_ring` instead once I observed https://github.com/oscar-system/Oscar.jl/issues/938.
- Use fmpz-valued rays of fans: https://github.com/oscar-system/Oscar.jl/pull/939/commits/192682550d33ef364182993eb34f41ef1da0995d. This uses the methods recently added by @benlorenz . In particular, this fixes the issue mentioned by @fieker in https://github.com/oscar-system/Oscar.jl/issues/932. Even more, I added this example as test in https://github.com/oscar-system/Oscar.jl/pull/939/commits/c989258ff8d118c1fedd30bc4be1659b29a6e865).
- Use simplicial complexes to compute the Stanley-Reisner ideal (as pointed out by @micjoswig): https://github.com/oscar-system/Oscar.jl/pull/939/commits/245fa43ac8030c61cc6f2036b65c47ed87404365, https://github.com/oscar-system/Oscar.jl/pull/939/commits/245fa43ac8030c61cc6f2036b65c47ed87404365. Note that there is currently no method to compute the polyhedron of a toric variety (if this exists). Rather, the only way to currently take advantage of this is to construct a variety from a polyhedron. The corresponding constructor then caches the polyhedron. I envision that this can be improved in a future PR, if people are ok with the current state.
- Following the recent improvements of `Base.show` by @lkastner, I extend the display method of toric varieties further to also display the coefficient ring: https://github.com/oscar-system/Oscar.jl/pull/939/commits/38105b9075af435679eeb88e20eba51dd52728b1
 
cc: @fingolfin and @thofma

thofma commented on src/Combinatorics/SimplicialComplexes.jl:
```diff
@@ -310,6 +310,22 @@ function stanley_reisner_ideal(R::MPolyRing, K::SimplicialComplex)
+
+Return the Stanley-Reisner ideal of the abstract simplicial complex `K` in the given ring `R`.
+This method is tailor-made to accept the graded Cox ring of a toric variety.
+
```
No new line or an example :)

HereAround replied:
Ok. I will adjust this upon resolving the points mentioned below (regarding two methods, integers as coefficients...). More below.

thofma commented:
There are now two methods
```
    stanley_reisner_ideal(R::MPolyRing, K::SimplicialComplex)

Return the Stanley-Reisner ideal of the abstract simplicial complex `K`, in the given ring `R`.
```
and
```
    stanley_reisner_ideal(R::MPolyRing{fmpq}, K::SimplicialComplex)

Return the Stanley-Reisner ideal of the abstract simplicial complex `K` in the given ring `R`.
This method is tailor-made to accept the graded Cox ring of a toric variety.
```
right? Does the docstring of the first method needs to be updated or do these two methods really have different assumptions on the input?

benlorenz commented on test/ToricVarieties/runtests.jl:
```diff
@@ -191,6 +191,13 @@ end
     @test is_projective_space(ntv2) == false
 end
 
+P = polarize(Polyhedron(Polymake.polytope.rand_sphere(5,60)))
```
```suggestion
P = polarize(Polyhedron(Polymake.polytope.rand_sphere(5,60; seed=42)))
```
Please add a seed here to make sure the testsuite is not affected by the random number generation.

HereAround replied:
Thank you @benlorenz !

benlorenz commented on src/ToricVarieties/NormalToricVarieties/attributes.jl:
```diff
@@ -95,10 +145,18 @@ julia> ngens(stanley_reisner_ideal(p2))
+        if has_attribute(v, :polyhedron)
+            # compute via simplicial complex
+            I = get_attribute(v, :polyhedron).pm_polytope.FACETS_THRU_VERTICES
+            K = SimplicialComplex(collect(Vector{Int}(Polymake.row(I,k)) for k = 1:size(I)[1]))
```
```suggestion
            I = pm_object(get_attribute(v, :polyhedron)).FACETS_THRU_VERTICES
            K = SimplicialComplex([Polymake.row(I,k) for k = 1:size(I)[1]])
```

micjoswig commented:
function stanley_reisner_ideal(R::MPolyRing{fmpq}, K::SimplicialComplex)

in

https://github.com/HereAround/Oscar.jl/blob/38105b9075af435679eeb88e20eba51dd52728b1/src/Combinatorics/SimplicialComplexes.jl#L321

Forcing rational coefficients here is unwanted! `QQ` is the default, and that is fine. But, for my purposes, I need `ZZ` sometimes.

HereAround commented:
> There are now two methods
> 
> ```
>     stanley_reisner_ideal(R::MPolyRing, K::SimplicialComplex)
> 
> Return the Stanley-Reisner ideal of the abstract simplicial complex `K`, in the given ring `R`.
> ```
> 
> and
> 
> ```
>     stanley_reisner_ideal(R::MPolyRing{fmpq}, K::SimplicialComplex)
> 
> Return the Stanley-Reisner ideal of the abstract simplicial complex `K` in the given ring `R`.
> This method is tailor-made to accept the graded Cox ring of a toric variety.
> ```
> 
> right? Does the docstring of the first method needs to be updated or do these two methods really have different assumptions on the input?

Maybe it is easiest to explain what was to be accomplished. And then people can tell if my proposal makes sense.

The original method for simplicial_complexes allows for `R::MPolyRing` as input. In extending to toric varieties, we also need to cover graded polynomial rings. Specifically, here are the data types of the (Cox) rings in question:

> julia> typeof(cox_ring(toric_projective_space(2)))
> MPolyRing_dec{fmpq, FmpqMPolyRing}

> julia> typeof(cox_ring(toric_projective_space(2,ZZ)))
> MPolyRing_dec{fmpz, FmpqMPolyRing}

In taking the demand of @micjoswig into account, I have therefore been trying to replace `R::MPolyRing` by a ring type, which covers all of the following:
1. Z^n-graded polynomial rings over Q and Z (or possibly even any finite field - but as of now Q and Z seem most important).
2. Non-graded polynomial rings over Q and Z (or possibly even any finite field - but again, as of now Q and Z seem most important).

I have not found one that does this. Rather, I covered those cases separately. For 2. is left the original method with `R::MPolyRing`. To cover 1 I looked at:

> supertype(typeof(cox_ring(toric_projective_space(2))))
MPolyRing{fmpq}
> supertype(typeof(cox_ring(toric_projective_space(2,ZZ))))
MPolyRing{fmpz}

and decided to use `MPolyRing{fmpq}`. Indeed, one can pass a ring with integer coefficients to this new method @micjoswig . As an example, consider the following code:

> P = polarize(Polyhedron(Polymake.polytope.rand_sphere(5,60)))
> v = NormalToricVariety(P,ZZ)
> I = Oscar.pm_object(get_attribute(v, :polyhedron)).FACETS_THRU_VERTICES
> K = SimplicialComplex([Polymake.row(I,k) for k = 1:size(I)[1]])
> stanley_reisner_ideal(cox_ring(v), K)

Based on this PR, this code first creates a toric variety with coefficient ring being `Z`. In particular, note that `typeof(cox_ring(v))` returns `MPolyRing_dec{fmpz, FmpzMPolyRing}`. Still, this code executes and computes the corresponding ideal. So it is possible to pass a graded ring with integer coefficients to this new method, or simply use the old method and pass a non-graded polynomial ring with integer coefficients to that instead.

With that being said, my first attempt was to simply copy the original method twice, once with `R::MPolyRing` and once with `MPolyRing{fmpq}` as filters for the input. It then turned out, that some functionality that is available for `R::MPolyRing` is not for `MPolyRing{fmpq}`. Hence, the implementations are a little different, but should otherwise execute the very same task.

Does that seem to make sense? As always - I am happy to adjust with better alternatives. Advice very much appreciated!

thofma commented:
OK, I must admit I still don't understand it, in particular by the distinction between `MPolyRing` or `MPolyRing{fmpq}`.

Do I understand correctly that specifying that I am working with a toric variety (and thus expecting some different behavior?) is done by changing the element type of the polynomial ring?

It seems to me that you want to distinguish between graded and non-graded polynomial rings. Why don't take `MPolyRing_dec` for the method where you want a graded ring?

lkastner commented on src/ToricVarieties/NormalToricVarieties/attributes.jl:
```diff
@@ -95,10 +145,18 @@ julia> ngens(stanley_reisner_ideal(p2))
+        if has_attribute(v, :polyhedron)
+            # compute via simplicial complex
+            I = pm_object(get_attribute(v, :polyhedron)).FACETS_THRU_VERTICES
+            K = SimplicialComplex([Polymake.row(I,k) for k = 1:size(I)[1]])
```
```suggestion
            I = pm_object(get_attribute(v, :polyhedron)).FACETS_THRU_VERTICES
            K = SimplicialComplex(I)
```
Does this maybe work? There is a constructor of `SimplicialComplex` that just takes an `IncidenceMatrix`.

HereAround commented:
Long story short - I think I simply got confused about the rings. Thank you @thofma for discussing earlier today.

While I am sure that I had an error at an earlier stage, now everything works with the currently implemented method for the stanley reisner ideal of a simplicial complex. Hence, I have removed my changes to the simplicial complexes.

lkastner commented:
I cannot find the place where it was agreed on that toric varieties should always carry a coefficient ring. Perhaps you can point me towards it? Sorry, but I have been rather busy lately with the workshop, etc.

HereAround commented:
The discussion started in https://github.com/oscar-system/Oscar.jl/issues/932. This change is tailor-made (or so I do hope) towards the request of @micjoswig .

lkastner commented:
Ok, but why not just give those methods that return some commutative algebra object a ring?

HereAround commented:
1. From the theory point of view, the coefficient ring is a datum of the variety. This also reflects designs for (projective) schemes.
2. I think we will eventually attach a lot of objects to these toric varieties which rely on the coefficient ring (defining equations for subvarieties, fp. graded modules for coherent sheaves). For this it seems efficient to access an attribute of the toric variety.

lkastner commented:
> 1. From the theory point of view, the coefficient ring is a datum of the variety. This also reflects designs for (projective) schemes.
> 
You have a point, nevertheless I think the toric geometry point of view is different, as it emphasises that everything can be computed from the combinatorics. Hence also most of the code is about the combinatorics and does not need the coefficient ring. I also wonder what algorithms actually do the right thing over arbitrary rings. 
- Why not make this optional via attribute system? I had a brief look at `M2` and there the coefficient ring is also just optional. By optional I really mean optional and not having a default value. The default value still forces one to take this into consideration in various places.
- I would also not consider the ring in the `Base.show` function. Currently the printout of the rings makes this result in grammatically weird sentences.
- For the methods returning rings and ideals it would still make sense to have the optional ring argument. It happens quite often that one starts working over `QQ` and then needs to look at something special happening over `ZZ`. Also the code for the Stanley-Reisner stuff then has a signature similar to the code for the same thing with the `SimplicialComplex`es.

>     2. I think we will eventually attach a lot of objects to these toric varieties which rely on the coefficient ring (defining equations for subvarieties, fp. graded modules for coherent sheaves). For this it seems efficient to access an attribute of the toric variety.

Yes, but the interesting objects will be torus invariant and then they will often just come from different combinatorial construction and the interesting algorithms are of combinatorial nature again. The toric divisors do not need the ring for example. Furthermore these objects could carry the ring instead of the variety. I would also say that it suffices to introduce this property once we create those objects.

Maybe @micjoswig can say what he has in mind?

lkastner commented on src/ToricVarieties/NormalToricVarieties/constructors.jl:
```diff
@@ -510,7 +576,7 @@ function Base.show(io::IO, v::AbstractNormalToricVariety)
 
     properties_string = [join(properties_string, ", ")]
-    push!(properties_string, "toric variety")
+    push!(properties_string, "toric variety over " * lowercase(string(coefficient_ring(v))))
```
I would really recommend not printing the coefficient ring. The result is awkward (in the examples I would expect "over the field of rational numbers" to be the correct thing). And with the `lowercase` we will also get things like `galois`.

HereAround replied:
As we discussed earlier this week, I have adjusted the print method such that for `QQ` and `ZZ` as coefficient ring, we print `QQ` and `ZZ`, respectively. In the other cases it prints `coefficient_ring(variety)`. Unfortunately, `string(variety)` returns the value of `Base.show` rather than the name of the variable which in the current session encodes the toric variety in question. But maybe this is acceptable?

lkastner commented on src/ToricVarieties/NormalToricVarieties/auxilliary.jl:
```diff
@@ -18,12 +18,12 @@ julia> A = [-1 -1 0 2; 2 3 -2 -1]
 function binomial_exponents_to_ideal(binoms::Union{AbstractMatrix, fmpz_mat})
     nvars = ncols(binoms)
-    R::FmpqMPolyRing, x = PolynomialRing(QQ, "x" => 1:nvars, cached=false)
+    R::FmpqMPolyRing, x = PolynomialRing(QQ, ["x" * string(i) for i in 1:nvars], cached=false)
```
Shouldn't this use the `coefficient_ring` then too?

HereAround replied:
Yes, I changed this.

lkastner reviewed (commented):
I am unsure why the change `x[1]->x1` is necessary. In Singular there used to be the notation that e.g. `x2` is a shorthand for `x^2`, so this might be confusing.

I am not sure whether we really should cache the heavier objects, like rings and ideals. For example the toric ideal can be computed very fast from the exponent vectors of the binomials. So I would suggest to just cache these and not the ideal itself, i.e. cache a matrix instead of an ideal. Would similar approaches work for the `stanley_reisner_*` and `cox_ring`?

micjoswig commented:
> Maybe @micjoswig can say what he has in mind?

An important reason for being able to specify the ring as an optional argument is that, in this way, users can pick variable names as they like.  Of course, variable names may be seen is eye candy altogether.  Yet, often a toric variety is produced from combinatorial data which comes with natural names.  And then we want to keep those names for suitable objects in the toric variety.  And the easiest way is to pass an entire ring as an additional argument.  By the way, this is in line of most of the commutative algebra code in OSCAR.

fingolfin commented:
This needs to be rebased

HereAround committed:
ToricVarieties: Compute SR-Ideal via simplicial complex if polytope of variety known

Co-authored-by: Benjamin Lorenz <benlorenz@users.noreply.github.com>

HereAround committed:
ToricVarieties: Initialize list of generators with correct type

HereAround committed:
ToricVarieties: Add set_coefficient_ring and coefficient_ring

HereAround committed:
ToricVarieties: Add set_coordinate_names and coordinate_names

HereAround committed:
ToricVarieties: Cox ring uses coordinate_names(v) as indeterminates

HereAround committed:
ToricVarieties: Cox ring uses coefficient_ring(v) as coefficient ring

HereAround committed:
ToricVarieties: Relax construction of projective_space and Hirzebruch_surface
-> Cox ring not set upon construction
-> User can choose coordinate_names and coefficient_ring

HereAround committed:
ToricVarieties: Relax construction of del_pezzo surfaces
-> Suggest coordinate names by setting the corresponding attribute
-> cox_ring not set so user can overwrite this choice and set coefficient_ring

HereAround committed:
ToricVarieties: toric_ideal uses coefficient_ring

HereAround committed:
ToricVarieties: Blowup_on_ith_minimal_torus_orbit uses coefficient_ring

HereAround committed:
ToricVarieties: Add test for set_coordinate_names and set_coefficient_ring

HereAround committed:
ToricVarieties: Extend documentation on default coefficient_ring and coordinate_names

HereAround committed:
ToricVarieties: More efficient simplex computation

Co-authored-by: Lars Kastner <lkastner@users.noreply.github.com>

HereAround commented:
@lkastner: I have updated this PR as we discussed earlier this week.

- There exists now an attribute `coefficient_ring` which can be set via `set_coefficient_ring` and can be accessed via `coefficient_ring(v)`.
- Likewise, one can set the homogeneous coordinates via `set_coordinate_names` and get their current value via `coordinate_names`.
- Both are NOT set upon construction.
- `coordinate_names` can be changed until `cox_ring` is computed for the first time. If no choice was made, I invoke the default choice `x1, x2, ...` and prevent changes from this point on.
- Likewise, `coefficient_ring` can be changed until either `cox_ring` or `toric_ideal` is computed for the first time. If no choice was made until then, the default `QQ` is invoked. After this point, a change of `coefficient_ring` is prevented.
- I think that freezing the value for these two attributes is natural at this point and that it prevents possible later inconsistencies. As always, I am happy to be proven wrong/corrected and to update with a better solution.
- If desired, I could add additional convenience methods to compute the ideals/ring with several other coefficient rings. Namely, I imagine a method which would take a ring as additional argument and then return the corresponding ring/ideal. These methods could then be used even after a choice of `coefficient_ring`/`coordinate_names` has been fixed. Let me know if this is desired.
- Your point regarding the caching is noted. I believe this PR is complicated already. I suggest to discuss this separately, e.g. in https://github.com/oscar-system/Oscar.jl/issues/997.
- The `base.show` method was also adjusted. Unless a value for `coefficient_ring` is known, none is (of course) printed. Once a value is known, it checks for `QQ` and `ZZ` (most prominent probably) and then prints "... toric variety over QQ" respectively "... toric variety over ZZ". For other rings, the print-out reads "...over coefficient_ring(variety)". Unfortunately, I could not obtain the value/string of the variable which represent the `toric_variety` in question in the active julia session via `string(v)`. The latter seems to return the same string as `base.show`. Hence, I wrote "variety" instead. I hope this is also acceptable.
- The simple command for the simplicial complex that you mentioned above seems to work. This was added in the last commit.

HereAround commented:
> This needs to be rebased

Done.

lkastner commented:
Great work, thank you. Let's discuss everything else after this is merged, since it is very large already as you said.

lkastner commented:
@fingolfin Should this then only be merged or squash+merge?

thofma commented:
Do the individual commits yield running versions of Oscar and do you want to preserve the individual commits? If yes, merge, otherwise, squash+merge.

HereAround commented:
My goal is to make sure that all commits yield a running version. However, due to the long runtime of the tests (in particular the documentation), I have not explicitly checked all commits, but only some during development. Feel free to squash if that seems more reasonable.
